File: internal/mcp/diff.go
package mcp

import (
	"regexp"
	"strings"
)

func ApplyUnifiedDiff(original, diff string) (string, error) {
	lines := strings.Split(original, "\n")
	diffLines := strings.Split(diff, "\n")

	hunks, err := parseHunks(diffLines)
	if err != nil {
		return "", err
	}

	for i := len(hunks) - 1; i >= 0; i-- {
		lines, err = applyHunk(lines, hunks[i])
		if err != nil {
			return "", err
		}
	}

	return strings.Join(lines, "\n"), nil
}

type hunk struct {
	oldStart int
	oldCount int
	newStart int
	newCount int
	changes  []diffLine
}

type diffLine struct {
	kind    byte
	content string
}

var hunkHeaderRegex = regexp.MustCompile(`^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@`)

func parseHunks(diffLines []string) ([]hunk, error) {
	var hunks []hunk
	var currentHunk *hunk

	for _, line := range diffLines {
		if strings.HasPrefix(line, "@@ ") {
			if currentHunk != nil {
				hunks = append(hunks, *currentHunk)
			}

			matches := hunkHeaderRegex.FindStringSubmatch(line)
			if matches == nil {
				continue
			}

			currentHunk = &hunk{}
			if matches[1] != "" {
				currentHunk.oldStart = parseInt(matches[1])
			}
			if matches[2] != "" {
				currentHunk.oldCount = parseInt(matches[2])
			} else {
				currentHunk.oldCount = 1
			}
			if matches[3] != "" {
				currentHunk.newStart = parseInt(matches[3])
			}
			if matches[4] != "" {
				currentHunk.newCount = parseInt(matches[4])
			} else {
				currentHunk.newCount = 1
			}
		} else if currentHunk != nil && len(line) > 0 {
			kind := line[0]
			if kind == '+' || kind == '-' || kind == ' ' {
				currentHunk.changes = append(currentHunk.changes, diffLine{
					kind:    kind,
					content: line[1:],
				})
			}
		}
	}

	if currentHunk != nil {
		hunks = append(hunks, *currentHunk)
	}

	return hunks, nil
}

func parseInt(s string) int {
	var result int
	for _, c := range strings.TrimSpace(s) {
		if c >= '0' && c <= '9' {
			result = result*10 + int(c-'0')
		}
	}
	return result
}

func applyHunk(lines []string, h hunk) ([]string, error) {
	var result []string
	lineIdx := 0

	for lineIdx < h.oldStart-1 && lineIdx < len(lines) {
		result = append(result, lines[lineIdx])
		lineIdx++
	}

	for _, change := range h.changes {
		switch change.kind {
		case ' ':
			if lineIdx < len(lines) {
				result = append(result, lines[lineIdx])
				lineIdx++
			}
		case '-':
			lineIdx++
		case '+':
			result = append(result, change.content)
		}
	}

	for lineIdx < len(lines) {
		result = append(result, lines[lineIdx])
		lineIdx++
	}

	return result, nil
}

File: reducto/diff.py
"""Unified diff parsing and application."""

from __future__ import annotations

import re

_HUNK = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")


class DiffError(Exception):
    pass


def apply_unified_diff(original: str, diff: str) -> str:
    lines = original.split("\n")
    hunks = _parse_hunks(diff.split("\n"))
    if not hunks and diff.strip():
        raise DiffError("No valid hunks found in diff")
    for hunk in reversed(hunks):
        lines = _apply_hunk(lines, hunk)
    return "\n".join(lines)


def _parse_hunks(diff_lines: list[str]) -> list[dict]:
    hunks: list[dict] = []
    current: dict | None = None
    for line in diff_lines:
        if line.startswith("@@ "):
            if current:
                hunks.append(current)
            m = _HUNK.match(line)
            if not m:
                continue
            current = {"old_start": int(m.group(1)), "changes": []}
        elif current and line and line[0] in "+- ":
            current["changes"].append((line[0], line[1:]))
    if current:
        hunks.append(current)
    return hunks


def _apply_hunk(lines: list[str], hunk: dict) -> list[str]:
    result: list[str] = []
    idx = 0
    while idx < hunk["old_start"] - 1 and idx < len(lines):
        result.append(lines[idx])
        idx += 1
    for kind, content in hunk["changes"]:
        if kind == " ":
            if idx < len(lines):
                result.append(lines[idx])
                idx += 1
        elif kind == "-":
            idx += 1
        elif kind == "+":
            result.append(content)
    while idx < len(lines):
        result.append(lines[idx])
        idx += 1
    return result
